// xpring-js/build/XRP/shared/account-set-flag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountSetFlag = void 0;
/**
 * Flags used in AccountSet transactions.
 *
 * @see https://xrpl.org/accountset.html#accountset-flags
 */
var AccountSetFlag;
(function (AccountSetFlag) {
    AccountSetFlag[AccountSetFlag["asfRequireDest"] = 1] = "asfRequireDest";
    AccountSetFlag[AccountSetFlag["asfRequireAuth"] = 2] = "asfRequireAuth";
    AccountSetFlag[AccountSetFlag["asfDisallowXRP"] = 3] = "asfDisallowXRP";
    AccountSetFlag[AccountSetFlag["asfDisableMaster"] = 4] = "asfDisableMaster";
    AccountSetFlag[AccountSetFlag["asfAccountTxnID"] = 5] = "asfAccountTxnID";
    AccountSetFlag[AccountSetFlag["asfNoFreeze"] = 6] = "asfNoFreeze";
    AccountSetFlag[AccountSetFlag["asfGlobalFreeze"] = 7] = "asfGlobalFreeze";
    AccountSetFlag[AccountSetFlag["asfDefaultRipple"] = 8] = "asfDefaultRipple";
    AccountSetFlag[AccountSetFlag["asfDepositAuth"] = 9] = "asfDepositAuth";
})(AccountSetFlag = exports.AccountSetFlag || (exports.AccountSetFlag = {}));
