// xpring-js/build/XRP/shared/gateway-balances.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.gatewayBalancesFromResponse = void 0;
const xpring_common_js_1 = require("xpring-common-js");
const _1 = require(".");
function gatewayBalancesFromResponse(gatewayBalancesResponse) {
    const { result } = gatewayBalancesResponse;
    if (!result.validated) {
        throw new _1.XrpError(_1.XrpErrorType.MalformedResponse, 'Gateway Balances response indicates unvalidated ledger.');
    }
    const account = result.account;
    if (!account) {
        throw new _1.XrpError(_1.XrpErrorType.MalformedResponse, 'gatewayBalancesResponse is missing required field `account`.');
    }
    const xAddress = xpring_common_js_1.XrpUtils.encodeXAddress(account, undefined, true);
    if (!xAddress) {
        throw new _1.XrpError(_1.XrpErrorType.MalformedResponse, 'Could not convert account to X-Address.');
    }
    return {
        account: xAddress,
        ledgerHash: result.ledger_hash,
        assets: result.assets,
        balances: result.balances,
        obligations: result.obligations,
    };
}
exports.gatewayBalancesFromResponse = gatewayBalancesFromResponse;
