// xpring-js/build/XRP/shared/transaction-status.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/** Represents statuses of transactions. */
var TransactionStatus;
(function (TransactionStatus) {
    /**
     * The transaction failed because the provided paths did not have enough liquidity to send anything at all.
     * This could mean that the source and destination accounts are not linked by trust lines.
     * The transaction cost was destroyed.
     */
    TransactionStatus[TransactionStatus["ClaimedCostOnly_PathDry"] = 0] = "ClaimedCostOnly_PathDry";
    /**
     * The transaction failed because the provided paths did not have enough liquidity to send the full amount.
     * The transaction cost was destroyed.
     */
    TransactionStatus[TransactionStatus["ClaimedCostOnly_PathPartial"] = 1] = "ClaimedCostOnly_PathPartial";
    /** The transaction did not achieve its intended purpose, but the transaction cost was destroyed. */
    TransactionStatus[TransactionStatus["ClaimedCostOnly"] = 2] = "ClaimedCostOnly";
    /** The transaction is not included in a finalized ledger, and is not valid, due to improper syntax, conflicting options, a bad signature, or something else. */
    TransactionStatus[TransactionStatus["MalformedTransaction"] = 3] = "MalformedTransaction";
    /** The transaction was included in a finalized ledger and failed. */
    TransactionStatus[TransactionStatus["Failed"] = 4] = "Failed";
    /** The transaction is not included in a finalized ledger. */
    TransactionStatus[TransactionStatus["Pending"] = 5] = "Pending";
    /** The transaction was included in a finalized ledger and succeeded. */
    TransactionStatus[TransactionStatus["Succeeded"] = 6] = "Succeeded";
    /** The transaction's last ledger sequence has been surpassed; the transaction will never be included in a validated ledger. */
    TransactionStatus[TransactionStatus["LastLedgerSequenceExpired"] = 7] = "LastLedgerSequenceExpired";
    /** The transaction status is unknown. */
    TransactionStatus[TransactionStatus["Unknown"] = 8] = "Unknown";
})(TransactionStatus || (TransactionStatus = {}));
exports.default = TransactionStatus;
