// xpring-js/build/XRP/shared/xrp-transaction-type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Types of transactions on the XRP Ledger.
 *
 * This is a partial list. Please file an issue if you have a use case that requires additional types.
 *
 * @see: https://xrpl.org/transaction-formats.html
 */
var XrpTransactionType;
(function (XrpTransactionType) {
    XrpTransactionType[XrpTransactionType["Payment"] = 0] = "Payment";
    XrpTransactionType[XrpTransactionType["AccountSet"] = 1] = "AccountSet";
    XrpTransactionType[XrpTransactionType["AccountDelete"] = 2] = "AccountDelete";
    XrpTransactionType[XrpTransactionType["CheckCancel"] = 3] = "CheckCancel";
    XrpTransactionType[XrpTransactionType["CheckCash"] = 4] = "CheckCash";
    XrpTransactionType[XrpTransactionType["CheckCreate"] = 5] = "CheckCreate";
    XrpTransactionType[XrpTransactionType["DepositPreauth"] = 6] = "DepositPreauth";
    XrpTransactionType[XrpTransactionType["EscrowCancel"] = 7] = "EscrowCancel";
    XrpTransactionType[XrpTransactionType["EscrowCreate"] = 8] = "EscrowCreate";
    XrpTransactionType[XrpTransactionType["EscrowFinish"] = 9] = "EscrowFinish";
    XrpTransactionType[XrpTransactionType["OfferCancel"] = 10] = "OfferCancel";
    XrpTransactionType[XrpTransactionType["OfferCreate"] = 11] = "OfferCreate";
    XrpTransactionType[XrpTransactionType["PaymentChannelClaim"] = 12] = "PaymentChannelClaim";
    XrpTransactionType[XrpTransactionType["PaymentChannelCreate"] = 13] = "PaymentChannelCreate";
    XrpTransactionType[XrpTransactionType["PaymentChannelFund"] = 14] = "PaymentChannelFund";
    XrpTransactionType[XrpTransactionType["SetRegularKey"] = 15] = "SetRegularKey";
    XrpTransactionType[XrpTransactionType["SignerListSet"] = 16] = "SignerListSet";
    XrpTransactionType[XrpTransactionType["TrustSet"] = 17] = "TrustSet";
})(XrpTransactionType || (XrpTransactionType = {}));
exports.default = XrpTransactionType;
