// xpring-js/build/ILP/model/account-balance.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountBalance = void 0;
const big_integer_1 = __importDefault(require("big-integer"));
/**
 * Response object for requests to get an account's balance
 */
class AccountBalance {
    /**
     * Private constructor to initialize an AccountBalance
     */
    constructor(options) {
        this.accountId = options.accountId;
        this.assetCode = options.assetCode;
        this.assetScale = options.assetScale;
        this.clearingBalance = options.clearingBalance;
        this.prepaidAmount = options.prepaidAmount;
        this.netBalance = options.clearingBalance.add(options.prepaidAmount);
    }
    /**
     * Constructs an AccountBalance from a GetBalanceResponse
     *
     * @param getBalanceResponse a GetBalanceResponse (protobuf object) whose field values will be used
     *                           to construct an AccountBalance
     * @return an AccountBalance with its fields set via the analogous protobuf fields.
     */
    static from(getBalanceResponse) {
        return new AccountBalance({
            accountId: getBalanceResponse.getAccountId(),
            assetCode: getBalanceResponse.getAssetCode(),
            assetScale: getBalanceResponse.getAssetScale(),
            clearingBalance: big_integer_1.default(getBalanceResponse.getClearingBalance()),
            prepaidAmount: big_integer_1.default(getBalanceResponse.getPrepaidAmount()),
        });
    }
}
exports.AccountBalance = AccountBalance;
exports.default = AccountBalance;
