// xpring-js/build/ILP/model/payment-result.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaymentResult = void 0;
const big_integer_1 = __importDefault(require("big-integer"));
/**
 * A response object containing details about a requested payment
 */
class PaymentResult {
    /**
     * Initializes a new PaymentResult
     */
    constructor(options) {
        this.originalAmount = options.originalAmount;
        this.amountDelivered = options.amountDelivered;
        this.amountSent = options.amountSent;
        this.successfulPayment = options.successfulPayment;
    }
    /**
     * Constructs a PaymentResult from a protobuf SendPaymentResponse
     *
     * @param protoResponse a SendPaymentResponse to be converted
     * @return a PaymentResult with fields populated using the analogous fields in the proto object
     */
    static from(protoResponse) {
        return new PaymentResult({
            originalAmount: big_integer_1.default(protoResponse.getOriginalAmount()),
            amountDelivered: big_integer_1.default(protoResponse.getAmountDelivered()),
            amountSent: big_integer_1.default(protoResponse.getAmountSent()),
            successfulPayment: protoResponse.getSuccessfulPayment(),
        });
    }
}
exports.PaymentResult = PaymentResult;
exports.default = PaymentResult;
