// xpring-js/build/XRP/network-clients/grpc-xrp-network-client.web.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const get_account_transaction_history_pb_1 = require("../Generated/web/org/xrpl/rpc/v1/get_account_transaction_history_pb");
const xrp_ledger_grpc_web_pb_1 = require("../Generated/web/org/xrpl/rpc/v1/xrp_ledger_grpc_web_pb");
const account_pb_1 = require("../Generated/web/org/xrpl/rpc/v1/account_pb");
const submit_pb_1 = require("../Generated/web/org/xrpl/rpc/v1/submit_pb");
const get_transaction_pb_1 = require("../Generated/web/org/xrpl/rpc/v1/get_transaction_pb");
const get_fee_pb_1 = require("../Generated/web/org/xrpl/rpc/v1/get_fee_pb");
const get_account_info_pb_1 = require("../Generated/web/org/xrpl/rpc/v1/get_account_info_pb");
const utils_1 = __importDefault(require("../../Common/utils"));
/**
 * A GRPC Based network client.
 */
class XrpGrpcNetworkClient {
    constructor(grpcURL) {
        if (utils_1.default()) {
            try {
                // This polyfill hack enables XMLHttpRequest on the global node.js state
                global.XMLHttpRequest = require('xhr2'); // eslint-disable-line
            }
            catch (_a) {
                // Swallow the error here for browsers
            }
        }
        this.grpcClient = new xrp_ledger_grpc_web_pb_1.XRPLedgerAPIServiceClient(grpcURL);
    }
    async getAccountInfo(request) {
        return new Promise((resolve, reject) => {
            this.grpcClient.getAccountInfo(request, {}, (error, response) => {
                if (error != null || response == null) {
                    reject(error);
                    return;
                }
                resolve(response);
            });
        });
    }
    async getFee(request) {
        return new Promise((resolve, reject) => {
            this.grpcClient.getFee(request, {}, (error, response) => {
                if (error != null || response == null) {
                    reject(error);
                    return;
                }
                resolve(response);
            });
        });
    }
    async getTransaction(request) {
        return new Promise((resolve, reject) => {
            this.grpcClient.getTransaction(request, {}, (error, response) => {
                if (error != null || response == null) {
                    reject(error);
                    return;
                }
                resolve(response);
            });
        });
    }
    async submitTransaction(request) {
        return new Promise((resolve, reject) => {
            this.grpcClient.submitTransaction(request, {}, (error, response) => {
                if (error != null || response == null) {
                    reject(error);
                    return;
                }
                resolve(response);
            });
        });
    }
    async getTransactionHistory(request) {
        return new Promise((resolve, reject) => {
            this.grpcClient.getAccountTransactionHistory(request, {}, (error, response) => {
                if (error != null || response == null) {
                    reject(error);
                    return;
                }
                resolve(response);
            });
        });
    }
    /* eslint-disable class-methods-use-this */
    AccountAddress() {
        return new account_pb_1.AccountAddress();
    }
    GetAccountInfoRequest() {
        return new get_account_info_pb_1.GetAccountInfoRequest();
    }
    GetTransactionRequest() {
        return new get_transaction_pb_1.GetTransactionRequest();
    }
    GetFeeRequest() {
        return new get_fee_pb_1.GetFeeRequest();
    }
    SubmitTransactionRequest() {
        return new submit_pb_1.SubmitTransactionRequest();
    }
    GetAccountTransactionHistoryRequest() {
        return new get_account_transaction_history_pb_1.GetAccountTransactionHistoryRequest();
    }
}
exports.default = XrpGrpcNetworkClient;
