// xpring-js/build/ILP/grpc-ilp-network-client.web.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const get_balance_request_pb_1 = require("./Generated/web/get_balance_request_pb");
const send_payment_request_pb_1 = require("./Generated/web/send_payment_request_pb");
const balance_service_grpc_web_pb_1 = require("./Generated/web/balance_service_grpc_web_pb");
const ilp_over_http_service_grpc_web_pb_1 = require("./Generated/web/ilp_over_http_service_grpc_web_pb");
const utils_1 = __importDefault(require("../Common/utils"));
const ilp_credentials_web_1 = __importDefault(require("./auth/ilp-credentials.web"));
class GrpcIlpNetworkClientWeb {
    constructor(grpcURL) {
        if (utils_1.default()) {
            try {
                // This polyfill hack enables XMLHttpRequest on the global node.js state
                global.XMLHttpRequest = require('xhr2'); // eslint-disable-line
            }
            catch (_a) {
                // Swallow the error here for browsers
            }
        }
        this.balanceClient = new balance_service_grpc_web_pb_1.BalanceServiceClient(grpcURL, null, {
            withCredentials: 'true',
        });
        this.paymentClient = new ilp_over_http_service_grpc_web_pb_1.IlpOverHttpServiceClient(grpcURL, null, {
            withCredentials: 'true',
        });
    }
    getBalance(request, accessToken) {
        return new Promise((resolve, reject) => {
            this.balanceClient.getBalance(request, ilp_credentials_web_1.default.build(accessToken), (error, response) => {
                if (error || !response) {
                    reject(error);
                    return;
                }
                resolve(response);
            });
        });
    }
    send(request, accessToken) {
        return new Promise((resolve, reject) => {
            this.paymentClient.sendMoney(request, ilp_credentials_web_1.default.build(accessToken), (error, response) => {
                if (error || !response) {
                    reject(error);
                    return;
                }
                resolve(response);
            });
        });
    }
    /* eslint-disable class-methods-use-this */
    SendPaymentRequest() {
        return new send_payment_request_pb_1.SendPaymentRequest();
    }
    GetBalanceRequest() {
        return new get_balance_request_pb_1.GetBalanceRequest();
    }
}
exports.default = GrpcIlpNetworkClientWeb;
